src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/ProjectConfigurationStateSynchronizer.cs

```diff
@@ -69,24 +68,81 @@ public void Dispose()
     }
     private async ValueTask ProcessBatchAsync(ImmutableArray<Work> items, CancellationToken token)
     {
-        foreach (var item in items.GetMostRecentUniqueItems(Comparer.Instance))
+        // Clear out our helper collections.
+        _resetProjectMap.Clear();
+        _indicesToSkip.Clear();
+
+        using var itemsToProcess = new PooledArrayBuilder<Work>(items.Length);
+        var index = 0;
+
+        foreach (var item in items)
+        {
+            if (token.IsCancellationRequested)
+            {
+                return;
+            }
+
+            if (item is ResetProject reset)
+            {
+                // We shouldn't ever get two resets for the same project, due to GetMostRecentUniqueItems, so the dictionary Add
+                // should always be safe
+                Debug.Assert(!_resetProjectMap.ContainsKey(reset.ProjectKey));
+
+                _resetProjectMap.Add(reset.ProjectKey, (reset, index));
+                itemsToProcess.Add(reset);
+            }
+            else if (item is AddProject add &&
+                _resetProjectMap.TryGetValue(add.ProjectKey, out var previousReset))
+            {
+                // We've already seen a Reset for this file path and now we have an Add, so let's convert to an Update
+                // and skip the Reset.
+                _indicesToSkip.Add(previousReset.index);
+
+                var update = new UpdateProject(previousReset.work.ProjectKey, add.ProjectInfo);
+                itemsToProcess.Add(update);
+                _resetProjectMap.Remove(add.ProjectKey);
+            }
+            else
+            {
+                itemsToProcess.Add(item);
+            }
+
+            index++;
+        }
+
+        // Now that we've got the real items we want to process, we remove the items we want to skip
+        // and then we can process the most recent unique items as normal
+        for (var i = _indicesToSkip.Count - 1; i >= 0; i--)
+        {
+            itemsToProcess.RemoveAt(_indicesToSkip[i]);
```

Comment:
The problem is that we want to skip indicies before collapsing. So for example, if we create a new bin file, then update it, this method will see an Add, a Remove, then an Add. We need to collapse that to Add then Update, and skip index 1. If we do that in the look that calls `GetMostRecentUniqueItems` though, we'll be processing a list that contains Remove then Add, because the first Add will have been de-duped. Skiping index 1 there is not what we want.

Your points about copying are totally fair though. I guess I can inline the logic from `GetMostRecentUniqueItems` into the the loop above this, and do it all in one backwards loop through the set.

Reply:
Okay this was a big BIG pain, so I haven't quite got there, but see what you think.

I did remove a hashset, and remove the most egregious array resizing that this RemoveAt method causes. I also fixed a bug in the previous code that I didn't realise I had :)

Also, and you might not like this, but I swapped from records to classes. We don't actually want value semantics for the work items, and in fact we don't get it anyway (RazorProjectInfo doesn't provide an equality comparer), and I'm not entirely convinced that the tests were even valid. They are reusing the same project info for all work items, but in the real world, for different project info instances, I think the generated GetHashCode would have returned different values, resulting in no actual de-duping at runtime, because the HashSet Contains checked wouldn't have even called the Equals method in the comparer.